### ISSUE TITLE
Download option for the whole "The Turing Way" PDF doument

### ISSUE THREAD

**Issue by KujaEx** [alan-turing-institute/the-turing-way#2211]
I saw on the [page](https://the-turing-way.netlify.app/welcome.html) that it is only possible to download the current page (subchapter) as pdf. It would be nice to have a download option for the whole (200> page?) document.

Also, if it's possible, to have a pdf which is not generated from the web pages, because then you have on the left or right side columns with contents you don't need in a document.

**Comment by sjvrijn** [comment]
jupyterbook has instructions on how to generate a PDF of the whole book, that should be a good place to start

https://jupyterbook.org/advanced/pdf.html?highlight=download

**Comment by zimolzak** [comment]
Thank you for the Jupyterbook link. In my view, these instructions should be easier to find. Or (better yet) there could be a PDF generated for each new release (or major/minor sub-release, etc.). I will be following these instructions "some day," but it will take significant time that I don't have at the moment.

Probably there are others in the same boat (want PDF but we'll all have to do the same steps in parallel: follow the link, `pip install pyppeteer`, possibly install more dependencies,  run the build, etc.). Then, there may be those who want PDF but won't even want to "just" follow the Jupyterbook instructions. I could be very wrong, & this could be a rare problem. Really, I speak only for myself.

I'm glad I found this issue because offline reading is a big plus for me. I truly do appreciate the link, because without it I would be 100% lost.

Thank you for all the work on this project!

**Comment by bsipocz** [comment]
This came up again recently on slack.

**Comment by kallewesterling** [comment]
Testing out the steps in [the link](
https://jupyterbook.org/advanced/pdf.html?highlight=download) above:

1. `pip install pyppeteer`
2. `cd the-turing-way/book`
3. `jupyter-book build website/ --builder pdfhtml`

In the latter, I run into some problems. Here's the output (abbreviated):

```
Running Jupyter-Book v0.13.0
Source Folder: /Users/kwesterling/Repositories/the-turing-way/book/website
Config Path: /Users/kwesterling/Repositories/the-turing-way/book/website/_config.yml
Output Path: /Users/kwesterling/Repositories/the-turing-way/book/website/_build/html
Running Sphinx v4.5.0
loading translations [en]... done
loading pickled environment... checking bibtex cache... up to date
done
myst v0.15.2: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_extensions=['colon_fence', 'dollarmath', 'linkify', 'substitution', 'tasklist'], dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], url_schemes=['mailto', 'http', 'https'], heading_anchors=None, heading_slug_func=None, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [singlehtml]: all documents
updating environment: 0 added, 1 changed, 0 removed
reading sources... [100%] community-handbook/local-build                        
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
assembling single document... reproducible-research/reproducible-research reproducible-research/overview reproducible-research/overview/overview-definitions reproducible-research/overview/overview-benefit reproducible-research/overview/overview-barriers reproducible-research/overview/overview-resources reproducible-research/open
. . .
done
writing... done
writing additional files... done
copying images... [100%] figures/road-to-reproducibility.svg                    ditor.png
copying static files... done
copying extra files... done
dumping object inventory... done
build succeeded.

The HTML page is in website/_build/html.
Finished generating HTML for book...
Converting book HTML into PDF...
[INFO] Starting Chromium download.
100%|███████████████████████████████████████████████████████████████████████████████████████████| 86.8M/86.8M [00:05<00:00, 15.4Mb/s]
[INFO] Beginning extraction
[INFO] Chromium extracted to: /Users/kwesterling/Library/Application Support/pyppeteer/local-chromium/588429
```

Then, the script causes an error:

```
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniconda/base/bin/jupyter-book", line 8, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/jupyter_book/cli/main.py", line 323, in build
    builder_specific_actions(
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/jupyter_book/cli/main.py", line 581, in builder_specific_actions
    html_to_pdf(output_path.joinpath("index.html"), path_pdf_output)
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/jupyter_book/pdf.py", line 31, in html_to_pdf
    asyncio.get_event_loop().run_until_complete(_html_to_pdf(html_file, pdf_file))
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/jupyter_book/pdf.py", line 50, in _html_to_pdf
    await page.goto(f"file:///{html_file}", {"waitUntil": ["networkidle2"]})
  File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.9/site-packages/pyppeteer/page.py", line 837, in goto
    raise error
pyppeteer.errors.TimeoutError: Navigation Timeout Exceeded: 30000 ms exceeded.
```

Even when trying to run it multiple times, it fails. It looks as though there's too short timeouts set (or something is wrong on my serving end of things).

**Comment by kallewesterling** [comment]
To be fair, the [`pypetteer` package](https://github.com/pyppeteer/pyppeteer?tab=readme-ov-file), which jupyterbook has built their docs on, has this notice from 2021:

> Attention: This repo is unmaintained and has been outside of minor changes for a long time. Please consider [playwright-python](https://github.com/microsoft/playwright-python) as an alternative.

**Comment by bsipocz** [comment]
yeap, the timeout has already an open issue for it https://github.com/executablebooks/jupyter-book/issues/1732, but I can't find that anyone pointed out  the fact that `pypetter` is practically deprecated/unmaintained. Would you open an issue for jupyterbook?

**Comment by kallewesterling** [comment]
I can do!

**Comment by bsipocz** [comment]
I'm checking whether it's a feature we need/would use for my day work, too, if yes the I can totally justify spending time on figuring out how to use `playwright-python` and upstream all the findings. 🤞

**Comment by kallewesterling** [comment]
That sounds good -- and much better than spending time building a solution here that includes mucking about in a package's source files to change up the timeout limits... which we totally could do, but it's not really a sustainable solution, in my mind.

**Comment by bsipocz** [comment]
nope, I certainly want to find/use a solution that works for all jupyterbook-based projects and not just for TTW.

**Comment by kallewesterling** [comment]
100%

**Comment by JimMadge** [comment]
We should check if the [JupyterBook 2 pdf output](https://next.jupyterbook.org/start/export-pdfs) works better for us.
